File: java/src/main/java/com/aurelion/connector/core/Registration.java
package com.aurelion.connector.core;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class Registration {

    private final Config cfg;

    public Registration(Config cfg) {
        this.cfg = cfg;
    }

    public void publish(String eventType, String instanceId, List<String> tags) {
        Map<String, Object> message = new LinkedHashMap<>();
        message.put("event_type", eventType);
        message.put("instance_id", instanceId);
        message.put("tags", tags);

        MqPublisher.publishJson(cfg, cfg.registryExchange(), "topic",
                eventType, message, null);
    }

    public void heartbeatLoop(String instanceId, List<String> tags, int intervalSeconds) {
        while (true) {
            publish("connector.heartbeat", instanceId, tags);
            try {
                Thread.sleep(intervalSeconds * 1000L);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            }
        }
    }
}

File: csharp/Core/Registration.cs
namespace Aurelion.Connector.Core;

public sealed class Registration
{
    private readonly ConnectorConfig _cfg;

    public Registration(ConnectorConfig cfg) => _cfg = cfg;

    public void Publish(string eventType, string instanceId, List<string> tags)
    {
        var message = new Dictionary<string, object?>
        {
            ["event_type"] = eventType,
            ["instance_id"] = instanceId,
            ["tags"] = tags,
        };

        MqPublisher.PublishJson(_cfg, _cfg.RegistryExchange, "topic",
            eventType, message, null);
    }

    public void HeartbeatLoop(string instanceId, List<string> tags, int intervalSeconds)
    {
        while (true)
        {
            Publish("connector.heartbeat", instanceId, tags);
            Thread.Sleep(intervalSeconds * 1000);
        }
    }
}
